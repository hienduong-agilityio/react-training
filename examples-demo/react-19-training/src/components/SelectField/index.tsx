// Libraries
import { JSX, memo, ReactNode, SelectHTMLAttributes } from 'react';
import classNames from 'classnames';

export interface ISelectFieldProps extends SelectHTMLAttributes<HTMLSelectElement> {
  // label: The label displayed above the select element.
  label?: string;
  // errorMessage: The error message to display if validation fails.
  errorMessage?: string;
  // customClasses: The custom CSS classes for styling the select element.
  customClasses?: string;
  // children: The elements to be rendered within the select dropdown.
  children: ReactNode;
}

const selectDefaultClasses: string =
  'border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow';
const selectErrorClasses: string = 'border-red-600';
const errorMessagesClasses: string = 'mt-1 text-sm text-red-600';
const labelClasses: string = 'mb-2 capitalize text-lg font-medium text-gray-700';

/**
 * SelectField component
 *
 * @returns {JSX.Element} - SelectField element.
 */
const SelectField = ({
  id = '',
  label = '',
  name = '',
  errorMessage = '',
  value = '',
  defaultValue = '',
  customClasses = '',
  children,
  onChange = () => {}
}: ISelectFieldProps): JSX.Element => {
  const selectFieldClasses: string = classNames(selectDefaultClasses, customClasses, {
    [selectErrorClasses]: errorMessage
  });

  return (
    <div className='flex flex-col'>
      {label && (
        <label htmlFor={id} className={labelClasses}>
          {label}
        </label>
      )}
      <select
        onChange={onChange}
        value={value}
        id={id}
        name={name}
        defaultValue={defaultValue}
        className={selectFieldClasses}
      >
        {children}
      </select>
      {errorMessage && <span className={errorMessagesClasses}>{errorMessage}</span>}
    </div>
  );
};

export default memo(SelectField);
