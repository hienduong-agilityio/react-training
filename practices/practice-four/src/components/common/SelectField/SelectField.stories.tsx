import type { Meta, StoryObj } from '@storybook/react';
import SelectField from './index';

const meta: Meta<typeof SelectField> = {
  title: 'Components/common/SelectField',
  component: SelectField,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs'],
  argTypes: {
    label: {
      description: 'The label displayed above the select element.',
      control: { type: 'text' },
      defaultValue: ''
    },
    name: {
      description: 'The name attribute for the select element.',
      control: { type: 'text' }
    },
    errorMessage: {
      description: 'The error message to display if validation fails.',
      control: { type: 'text' },
      defaultValue: ''
    },
    defaultValue: {
      description: 'The default selected value of the select element.',
      control: { type: 'text' },
      defaultValue: ''
    },
    customClasses: {
      description: 'The custom CSS classes for styling the select element.',
      control: { type: 'text' },
      defaultValue: ''
    },
    children: {
      description: 'The elements to be rendered within the select dropdown.',
      control: { type: 'text' }
    },
    onChange: {
      description: 'Function to handle change events on the select element.',
      action: 'value changed'
    }
  }
};

export default meta;

type Story = StoryObj<typeof meta>;

// Default story with no error message
export const Default: Story = {
  args: {
    label: 'Select an option',
    name: 'example',
    children: (
      <>
        <option value='1'>Option 1</option>
        <option value='2'>Option 2</option>
        <option value='3'>Option 3</option>
      </>
    )
  }
};

// Story demonstrating an error message
export const WithErrorMessage: Story = {
  args: {
    label: 'Select an option',
    name: 'example',
    errorMessage: 'This field is required.',
    children: (
      <>
        <option value=''>Select an option...</option>
        <option value='1'>Option 1</option>
        <option value='2'>Option 2</option>
        <option value='3'>Option 3</option>
      </>
    )
  }
};

// Story with a default value set
export const WithDefaultValue: Story = {
  args: {
    label: 'Select an option',
    name: 'example',
    defaultValue: '2',
    children: (
      <>
        <option value='1'>Option 1</option>
        <option value='2'>Option 2</option>
        <option value='3'>Option 3</option>
      </>
    )
  }
};
