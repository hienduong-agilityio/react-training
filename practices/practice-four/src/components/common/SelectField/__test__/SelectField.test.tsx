// Libraries
import { render, screen, fireEvent } from '@testing-library/react';

// Components
import SelectField, { ISelectFieldProps } from '@/components/common/SelectField';

/*
SelectField Component
  √ should match the snapshot when rendered
  √ should render the select element with children
  √ should have an empty name attribute if not provided
  √ should handle the default onChange function without error if not provided
  √ should render with a label if provided and associate it with the select element
  √ should render the error message if provided
  √ should apply custom classes to the select element
  √ should render with the default value
 */
describe('SelectField Component', () => {
  const onChangeMock = jest.fn();
  let renderComponent: (props?: Partial<ISelectFieldProps>) => ReturnType<typeof render>;

  beforeEach(() => {
    jest.clearAllMocks();

    renderComponent = (props: Partial<ISelectFieldProps> = {}) => {
      return render(
        <SelectField name='test-select' onChange={onChangeMock} {...props}>
          <option value='1'>Option 1</option>
          <option value='2'>Option 2</option>
        </SelectField>
      );
    };
  });

  it('should match the snapshot when rendered', () => {
    const { container } = renderComponent({ label: 'Select an option', defaultValue: '1' });
    expect(container).toMatchSnapshot();
  });

  it('should render the select element with children', () => {
    renderComponent();
    const selectElement = screen.getByRole('combobox') as HTMLSelectElement;
    expect(selectElement).toBeInTheDocument();
    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Option 2')).toBeInTheDocument();

    const errorMessage = screen.queryByText('This field is required');
    expect(errorMessage).toBeNull();

    fireEvent.change(selectElement, { target: { value: '2' } });
    expect(onChangeMock).toHaveBeenCalledTimes(1);
  });

  it('should have an empty name attribute if not provided', () => {
    renderComponent({ name: '' });

    const selectElement = screen.getByRole('combobox') as HTMLSelectElement;
    expect(selectElement.name).toBe('');
  });

  it('should handle the default onChange function without error if not provided', () => {
    renderComponent({ onChange: undefined });
    const selectElement = screen.getByRole('combobox') as HTMLSelectElement;

    expect(() => {
      fireEvent.change(selectElement, { target: { value: '2' } });
    }).not.toThrow();
  });

  it('should render with a label if provided and associate it with the select element', () => {
    renderComponent({ id: 'select-test', label: 'Select an option' });

    const labelElement = screen.getByText('Select an option');
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toHaveAttribute('for', 'select-test');
  });

  it('should render the error message if provided', () => {
    renderComponent({ errorMessage: 'This field is required' });

    const errorMessage = screen.getByText('This field is required');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveClass('text-red-600');
  });

  it('should apply custom classes to the select element', () => {
    const customClass = 'custom-select-class';
    renderComponent({ customClasses: customClass });

    const selectElement = screen.getByRole('combobox') as HTMLSelectElement;
    expect(selectElement).toHaveClass(customClass);
  });

  it('should render with the default value', () => {
    renderComponent({ defaultValue: '2' });

    const selectElement = screen.getByRole('combobox') as HTMLSelectElement;
    expect(selectElement.value).toBe('1');
  });
});
