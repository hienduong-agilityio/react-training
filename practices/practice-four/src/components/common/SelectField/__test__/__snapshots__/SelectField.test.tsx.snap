// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField Component should match the snapshot when rendered 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <label
      class="mb-2 capitalize text-lg font-medium text-gray-700"
      for=""
    >
      Select an option
    </label>
    <select
      class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow"
      id=""
      name="test-select"
    >
      <option
        selected=""
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
    </select>
  </div>
</div>
`;
